consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,863 @@
+		cursor += 4
+		if len(data[cursor:]) < 1 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: PublicKey")
+		}
+		cursor++
+		if data[cursor-1] != nilMarker {
+			if len(data[cursor:]) < fieldparams.BLSPubkeyLength {
+				return nil, errors.Wrap(errDataSmall, "validatorDiffs: PublicKey")
+			}
+			validatorDiffs[i].PublicKey = data[cursor : cursor+fieldparams.BLSPubkeyLength]
+			cursor += fieldparams.BLSPubkeyLength
+		}
+		if len(data[cursor:]) < 1 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: WithdrawalCredentials")
+		}
+		cursor++
+		if data[cursor-1] != nilMarker {
+			if len(data[cursor:]) < fieldparams.RootLength {
+				return nil, errors.Wrap(errDataSmall, "validatorDiffs: WithdrawalCredentials")
+			}
+			validatorDiffs[i].WithdrawalCredentials = data[cursor : cursor+fieldparams.RootLength]
+			cursor += fieldparams.RootLength
+		}
+		if len(data[cursor:]) < 8 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: EffectiveBalance")
+		}
+		validatorDiffs[i].EffectiveBalance = binary.LittleEndian.Uint64(data[cursor : cursor+8])
+		cursor += 8
+		if len(data[cursor:]) < 1 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: Slashed")
+		}
+		validatorDiffs[i].Slashed = data[cursor] != nilMarker
+		cursor++
+		if len(data[cursor:]) < 8 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: ActivationEligibilityEpoch")
+		}
+		validatorDiffs[i].ActivationEligibilityEpoch = primitives.Epoch(binary.LittleEndian.Uint64(data[cursor : cursor+8]))
+		cursor += 8
+		if len(data[cursor:]) < 8 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: ActivationEpoch")
+		}
+		validatorDiffs[i].ActivationEpoch = primitives.Epoch(binary.LittleEndian.Uint64(data[cursor : cursor+8]))
+		cursor += 8
+		if len(data[cursor:]) < 8 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: ExitEpoch")
+		}
+		validatorDiffs[i].ExitEpoch = primitives.Epoch(binary.LittleEndian.Uint64(data[cursor : cursor+8]))
+		cursor += 8
+		if len(data[cursor:]) < 8 {
+			return nil, errors.Wrap(errDataSmall, "validatorDiffs: WithdrawableEpoch")
+		}
+		validatorDiffs[i].WithdrawableEpoch = primitives.Epoch(binary.LittleEndian.Uint64(data[cursor : cursor+8]))
+		cursor += 8
+	}
+	if cursor != len(data) {
+		return nil, errors.Errorf("data is too large, expected %d bytes, got %d", len(data), cursor)
+	}
+	return validatorDiffs, nil
+}
+
+// newBalancesDiff deserializes a new balances diff from the given data.
+func newBalancesDiff(data []byte) ([]int64, error) {
+	if len(data) < 8 {
+		return nil, errors.Wrap(errDataSmall, "balancesDiff")
+	}
+	balancesLength := binary.LittleEndian.Uint64(data[:8])
+	if len(data) != int(8+balancesLength*8) {
+		return nil, errors.Errorf("incorrect length of balancesDiff, expected %d, got %d", 8+balancesLength*8, len(data))
+	}
+	balances := make([]int64, balancesLength)
+	for i := range balancesLength {
+		balances[i] = int64(binary.LittleEndian.Uint64(data[8*(i*1) : 8*(i*2)]))
```

Comment:
I dont think this is right.
`i=0`  -> `[0:0]`
`i=1` -> `[8:16]`
`i=2` -> `[16:32]`
`i=3` -> `[24:48]`